src/librbd/image/CloneRequest.cc

```diff
@@ -0,0 +1,591 @@
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't set metadata: " << cpp_strerror(*r) << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else {
+    get_mirror_mode();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::get_mirror_mode() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  if (!m_imctx->test_features(RBD_FEATURE_JOURNALING)) {
+    send_close();
+    return;
+  }
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_mode_get_start(&op);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_callback<klass, &klass::handle_get_mirror_mode>(this);
+  m_out_bl.clear();
+  m_imctx->md_ctx.aio_operate(RBD_MIRRORING,
+			      comp, &op, &m_out_bl);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_get_mirror_mode(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r == 0) {
+    lderr(m_cct) << "failed to retrieve mirror mode: " << cpp_strerror(*r)
+	       << dendl;
+    bufferlist::iterator it = m_out_bl.begin();
+    *r = cls_client::mirror_mode_get_finish(&it, &m_mirror_mode);
+  }
+
+  if (*r < 0 && *r != -ENOENT) {
+    m_r_saved = *r;
+    send_remove_child();
+  } else {
+    if (m_mirror_mode == cls::rbd::MIRROR_MODE_POOL ||
+	!m_non_primary_global_image_id.empty()) {
+      send_enable_mirror();
+    } else {
+      send_close();
+    }
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_enable_mirror() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_enable_mirror>(this);
+
+  mirror::EnableRequest<I> *req = mirror::EnableRequest<I>::create(
+    m_imctx->md_ctx, m_id, m_non_primary_global_image_id,
+    m_imctx->op_work_queue, ctx);
+  req->send();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_enable_mirror(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "failed to enable mirroring: " << cpp_strerror(*r)
+               << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else {
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_close() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  assert(m_imctx != nullptr);
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_async_context_callback(
+    *m_imctx, create_context_callback<
+      klass, &klass::handle_close, false>(this));
+  m_imctx->state->close(ctx);
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_close(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_imctx->destroy();
+  m_imctx = nullptr;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't close image: " << cpp_strerror(*r) << dendl;
+    return m_on_finish;
+  }
+
+  if (m_r_saved == 0) {
+    complete(0);
```

Comment:
@yangdongsheng One of many examples of how ```complete(XYZ)``` is being used instead of just returning ```m_on_finish``` from the ```Context *handle_XYZ``` function.